pydrofoil/real.py

```diff
@@ -153,5 +167,57 @@ def toint(self):
     def totuple(self):
         return self.num.toint(), self.den.toint()
     
-    def sqrt(self):
-        return Real(self.num, self.den)
+    def sqrt(self): 
+        if self.num.int_lt(0):
+            assert False, "sqrt(x), x cannot be negative"
+        if self.num.int_eq(0):
+            return Real(NULLRBIGINT, ONERBIGINT)
+        OUT, PerfectSqr = isperfectsquare(self.num)
+        if PerfectSqr and self.den.int_eq(1):
+            return Real(OUT, self.den)
+        current = Real(isqrt(self.num), isqrt(self.den))
```

Comment:
this all looks good! I would like to change the detection of squares though. I don't understand the `iperfectsquare` algorithm and I think we can just delete it and write this instead (untested):

```python
current = Real(isqrt(self.num), isqrt(self.den))
if current.mul(current).eq(self):
    return current
```

apart from that, well done! will merge once this is fixed and CI passes.